File: Codeforces/segmentedSieve.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <algorithm>
using namespace std;


void sieve(int n, vector<int>& prime)
{
	bool mark[n+1] = {false};
	for (int p = 2; p*p < n; ++p)
	{
		if(mark[p] == false)
		{
			for(int i = 2*p; i < n; i += p)
			{
				mark[i] = true;
			}
		}
	}
	for(int p = 2; p <= n; p++)
	{
		if(mark[p] == false)
		{
			cout<<p<<" ";
			prime.push_back(p);
		}
	}
}

void segmentedSieve(int n)
{
	int limit = sqrt(n) + 1;
	vector<int> prime;
	sieve(limit, prime);
	int low = limit;
	int high = 2*limit;
	while(low < n)
	{
		bool mark[limit+1] = {};
		for(int i = 0; i < prime.size(); i++)
		{
			int lowLim = (low/prime[i])*prime[i];
			if(lowLim < low)
				lowLim += prime[i];
			for(int j = lowLim; j < high; j += prime[i])
			{
				mark[j-low] = true;
			}	
		}

		for(int i = low; i < high; i++)
		{
			if(mark[i-low] == false)
				cout<<i<<" ";
		}
		low += limit;
		high += limit;
		if(high > n)
			high = n;
	}
}
int main()
{
    int n = 100;
    segmentedSieve(n);
    return 0;
}

File: Codeforces/segmentedSieve.py
import math
def sieve(n):
	prime = list()
	mark = [True]*(n+1)
	p = 2
	while p*p <= n:
		if(mark[p]== True):
			for j in range(2*p,n+1,p):
				mark[j] = False
		p += 1
	for p in range(2,n+1):
		if(mark[p] == True):
			prime.append(p)
	return prime

def segmentedSieve(n):
	limit = math.floor(math.sqrt(n)) + 1
	prime = sieve(limit)
	low = limit
	high = 2*limit 
	for x in prime:
		print(x, end = ' ')
	while low < n:
		mark = [True]*(limit+1)
		for i in range(len(prime)):
			start = math.floor(low/prime[i])*prime[i]
			if(start < low):
				start += prime[i]
			for j in range(start,high,prime[i]):
				mark[j-low] = False

		for i in range(low,high):
			if(mark[i - low] == True):
				print(i, end = " ")

		low += limit
		high += limit
		if high > n:
			high = n

segmentedSieve(100)

